lib/ueberauth/strategies/helpers.ex

```diff
@@ -192,13 +211,23 @@ defmodule Ueberauth.Strategy.Helpers do
           Plug.Conn.t()
         ) :: keyword()
   def with_state_param(opts, conn) do
-    state = conn.private[:ueberauth_state_param]
+    state = %{}
 
-    if is_nil(state) do
-      opts
-    else
-      Keyword.put(opts, :state, state)
-    end
+    state =
+      if conn.private[:ueberauth_anti_csrf_token_state_param] do
+        Map.put(state, "csrf", conn.private[:ueberauth_anti_csrf_token_state_param])
+      else
+        state
+      end
+
+    state =
+      if conn.private[:ueberauth_state_param] do
+        Map.put(state, "data", conn.private[:ueberauth_state_param])
+      else
+        state
+      end
+
+    Keyword.put(opts, :state, Ueberauth.json_library().encode!(state))
```

Comment:
What would be the reason for encoding the `state` params as JSON instead of expecting a string?

Reply:
How would you pass the `csrf` and custom `data` field into the single `state` parameter that need to be send to oauth server?

Reply:
You would hash the value somehow or marshal it into a string format and use that.

Your system is the one that is aware of JSON format for such things, maybe you want to use some binary format that, or whatever other thing that is not JSON

Reply:
But I want to pass through Provider exact value of CSRF token (which we need to compare with token stored in session) as well as some custom data that client application want to read after callback.

So the whole query string `state` parameter passed to provider could look like:
`"{\"csrf\":\"csrf_token\",\"data\":\"any_custom_data\"}"`

Both coding and decoding is on the library side, so the application doesn't even know if it's JSON.